sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -10,14 +10,35 @@
 
 namespace Pulumi
 {
-    internal static class ResourcePackages
+    internal class ResourcePackages
```

Comment:
I'm somewhat confused why you needed to make all the changes related to `ResourcePackages` in this PR. Why can't it be static? Are we assuming that multiple deployments may run from different app domains? Why do you need to introduce an external resource test project? Why add the `resourcePackageAssemblies` parameter to `RunAsync`?

Reply:
The external assembly support is not relevant for console app execution, because with console app execution the executing assembly will always reference the necessary resource package assemblies so it will always be OK to fall back on the on `AppDomain.CurrentDomain`. So strictly speaking, it didn't *need* to be in this PR.

But, in that vein, concurrent execution isn't necessary for console execution either.

In any scenario where you actually might *need* to make use of concurrent execution, it is important that the executing assembly be able to explicitly pass in assemblies to be used as resource package assemblies. Because:
1. Individual deployments might use different resource package assemblies
2. The executing assembly (like a theoretical Web API) might not explicitly reference those resource package assemblies itself so `AppDomain.CurrentDomain` wouldn't find them
3. Those assemblies might be loaded at runtime

At my company we will be using an internal Web API to automate deployment, and our pulumi deployments are abstracted as plugins. There's a plugin for an "application type" i.e. a plugin to deploy a static website, or one of our larger proprietary application types. So, the Web API is going to load the plugin to get the pulumi deployment code. It will not itself have the resource package assemblies in its `AppDomain.CurrentDomain` necessary for the deployment to function as needed.

And since it is possible for those assemblies to be loaded at runtime, Pulumi should be sure not to cache any of them too. Because if a reference is held to an `Assembly` or `Type` than that assembly can never be un-loaded.

Reply:
Thank you for the great comment! I'm surprised you have put much thought into this functionality which is brand-new and not officially released yet (it's only needed for multi-language components).

Yet, without judging it as right or wrong, could you please split this change out of this PR?

Reply:
I'm confused- how would splitting this change out look? Is the point of this PR not to move away from static global state so pulumi can support concurrency? If this PR were to leave some static global state un-accounted for than it would be incomplete, no?

Reply:
> it would be incomplete, no?

IMO, it doesn't have to be complete - it should move a needle and fix a specific problem. For instance, the problem of not being able to run tests in parallel.

As of now, is there an observable behavior for this (ResourcePackages) global state, or a limitation that it brings?

Reply:
> For instance, the problem of not being able to run tests in parallel.

For what it's worth, I still was not able to enable test collection parallelization after the `AsyncLocal<T>` change to `Deployment.Instance`. After doing so I was encountering intermittent issues when running tests in parallel from the pulumi CLI, unrelated to anything in the dotnet SDK. It was trying to grab multiple file locks on my local pulumi `credentials.json` at once.

Back to topic though - Is the fact that it would be unable to function with runtime assemblies or assemblies not referenced by the executing assembly not a limitation?

Reply:
> After doing so I was encountering intermittent issues when running tests in parallel from the pulumi CLI, unrelated to anything in the dotnet SDK. 

This sounds to be specific to the automation API. Are you able to enable parallel tests in this PR?

> assemblies not referenced by the executing assembly not a limitation?

I'm fine with this for now, TBH. My goal is to have the smallest PR possible that still delivers value. If we need more PRs later, that's okay.

Reply:
OK, I have backtracked the resource package changes. Yes, you can enable test parallelization in this PR.